File: plugins/plugin-native-browser-surface/android/src/main/java/ai/eliza/plugins/browsersurface/BrowserSurfacePlugin.kt
package ai.eliza.plugins.browsersurface

import android.os.Build
import android.view.View
import android.webkit.WebView
import android.widget.FrameLayout
import androidx.webkit.ProfileStore
import androidx.webkit.WebViewCompat
import androidx.webkit.WebViewFeature
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

/**
 * Native Android half of `ElizaSurfaceManager` (#15245): layers one [WebView]
 * per Browser tab above the Capacitor host webview, each with the platform
 * out-of-process renderer and its OWN storage partition, so third-party content
 * can never reach the host realm or a sibling tab.
 *
 * Isolation maps onto two androidx.webkit primitives. Renderer: the WebView
 * renderer runs out-of-process by platform default on API 26+; an `isolated`
 * surface asserts that separation is actually in effect and fails fast if not.
 * Storage: an `isolated` surface gets its own multi-profile [androidx.webkit
 * Profile][ProfileStore] (cookies/localStorage/IndexedDB partitioned); a
 * `shared` surface uses the default profile. There is NO silent degrade — if the
 * system WebView is too old for multi-profile, `createSurface` rejects, because a
 * surface that quietly shares the default store is the exact leak this closes.
 */
@CapacitorPlugin(name = "ElizaSurfaceManager")
class ElizaSurfaceManagerPlugin : Plugin() {
    private data class Surface(
        val webView: WebView,
        val process: String,
        val storage: String,
        var foregrounded: Boolean,
    )

    private val surfaces = HashMap<String, Surface>()

    private fun density(): Float = activity.resources.displayMetrics.density

    @PluginMethod
    fun createSurface(call: PluginCall) {
        val id = call.getString("id") ?: run {
            call.reject("createSurface requires an id")
            return
        }
        val process = call.getString("process")
        if (process != "isolated" && process != "shared") {
            call.reject("createSurface requires an explicit process policy (isolated|shared)")
            return
        }
        val storage = call.getString("storage")
        if (storage != "isolated" && storage != "shared") {
            call.reject("createSurface requires an explicit storage policy (isolated|shared)")
            return
        }
        val url = call.getString("url")

        activity.runOnUiThread {
            if (surfaces.containsKey(id)) {
                call.resolve()
                return@runOnUiThread
            }
            val host = bridge.webView.parent as? FrameLayout ?: run {
                call.reject("host webview has no FrameLayout parent to attach the surface to")
                return@runOnUiThread
            }

            val webView = WebView(activity)
            webView.settings.javaScriptEnabled = true
            webView.settings.domStorageEnabled = true
            webView.settings.databaseEnabled = true

            // Storage isolation via multi-profile. Fail-fast: no silent degrade
            // to the shared default profile on an unsupported system WebView.
            if (storage == "isolated") {
                if (!WebViewFeature.isFeatureSupported(WebViewFeature.MULTI_PROFILE)) {
                    webView.destroy()
                    call.reject("isolated storage requires WebView multi-profile support; system WebView is too old")
                    return@runOnUiThread
                }
                val profile = ProfileStore.getInstance().getOrCreateProfile("eliza-surface-$id")
                WebViewCompat.setProfile(webView, profile.name)
            }
            // shared storage ⇒ the default profile (host-scoped store).

            val lp = FrameLayout.LayoutParams(0, 0)
            host.addView(webView, lp)
            webView.visibility = View.GONE
            if (url != null) webView.loadUrl(url)

            // Renderer isolation: assert the out-of-process renderer is in effect
            // for an isolated surface. On API 26+ the platform runs it in the
            // sandboxed :webview_service process; a null handle when the feature
            // is supported means this device/build cannot isolate — reject.
            if (
                process == "isolated" &&
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.O &&
                WebViewFeature.isFeatureSupported(WebViewFeature.GET_WEB_VIEW_RENDERER) &&
                WebViewCompat.getWebViewRenderProcess(webView) == null
            ) {
                host.removeView(webView)
                webView.destroy()
                call.reject("isolated process requires an out-of-process WebView renderer, which is unavailable on this device")
                return@runOnUiThread
            }

            surfaces[id] = Surface(webView, process, storage, false)
            call.resolve()
        }
    }

    @PluginMethod
    fun setBounds(call: PluginCall) {
        val id = call.getString("id") ?: run {
            call.reject("setBounds requires an id")
            return
        }
        val x = call.getDouble("x") ?: 0.0
        val y = call.getDouble("y") ?: 0.0
        val width = call.getDouble("width") ?: 0.0
        val height = call.getDouble("height") ?: 0.0
        activity.runOnUiThread {
            val surface = surfaces[id] ?: run {
                call.reject("no surface $id")
                return@runOnUiThread
            }
            val d = density()
            val lp = FrameLayout.LayoutParams((width * d).toInt(), (height * d).toInt())
            lp.leftMargin = (x * d).toInt()
            lp.topMargin = (y * d).toInt()
            surface.webView.layoutParams = lp
            call.resolve()
        }
    }

    @PluginMethod
    fun navigate(call: PluginCall) {
        val id = call.getString("id")
        val url = call.getString("url")
        if (id == null || url == null) {
            call.reject("navigate requires an id and a url")
            return
        }
        activity.runOnUiThread {
            val surface = surfaces[id] ?: run {
                call.reject("no surface $id")
                return@runOnUiThread
            }
            surface.webView.loadUrl(url)
            call.resolve()
        }
    }

    @PluginMethod
    fun foregroundSurface(call: PluginCall) {
        val id = call.getString("id") ?: run {
            call.reject("foregroundSurface requires an id")
            return
        }
        activity.runOnUiThread {
            val surface = surfaces[id] ?: run {
                call.reject("no surface $id")
                return@runOnUiThread
            }
            surface.webView.bringToFront()
            surface.webView.visibility = View.VISIBLE
            surface.foregrounded = true
            call.resolve()
        }
    }

    @PluginMethod
    fun backgroundSurface(call: PluginCall) {
        val id = call.getString("id") ?: run {
            call.reject("backgroundSurface requires an id")
            return
        }
        activity.runOnUiThread {
            val surface = surfaces[id] ?: run {
                call.reject("no surface $id")
                return@runOnUiThread
            }
            surface.webView.visibility = View.GONE
            surface.foregrounded = false
            call.resolve()
        }
    }

    @PluginMethod
    fun destroySurface(call: PluginCall) {
        val id = call.getString("id") ?: run {
            call.reject("destroySurface requires an id")
            return
        }
        activity.runOnUiThread {
            surfaces.remove(id)?.let { surface ->
                surface.webView.stopLoading()
                (surface.webView.parent as? FrameLayout)?.removeView(surface.webView)
                surface.webView.destroy()
            }
            call.resolve()
        }
    }

    @PluginMethod
    fun foregroundHost(call: PluginCall) {
        activity.runOnUiThread {
            for (surface in surfaces.values) {
                surface.webView.visibility = View.GONE
                surface.foregrounded = false
            }
            call.resolve()
        }
    }

    @PluginMethod
    fun getSurfaceState(call: PluginCall) {
        val id = call.getString("id") ?: run {
            call.reject("getSurfaceState requires an id")
            return
        }
        activity.runOnUiThread {
            val result = JSObject()
            val surface = surfaces[id]
            if (surface == null) {
                result.put("exists", false)
                result.put("foregrounded", false)
                result.put("currentUrl", JSObject.NULL)
                result.put("process", JSObject.NULL)
                result.put("storage", JSObject.NULL)
            } else {
                result.put("exists", true)
                result.put("foregrounded", surface.foregrounded)
                result.put("currentUrl", surface.webView.url ?: JSObject.NULL)
                result.put("process", surface.process)
                result.put("storage", surface.storage)
            }
            call.resolve(result)
        }
    }
}

File: plugins/plugin-native-browser-surface/ios/Sources/BrowserSurfacePlugin/BrowserSurfacePlugin.swift
import Foundation
import Capacitor
import WebKit
import UIKit

/// Native iOS half of `ElizaSurfaceManager` (#15245): layers one `WKWebView` per
/// Browser tab above the Capacitor host webview, each in its OWN renderer
/// process and storage partition, so third-party content can never reach the
/// host realm or a sibling tab.
///
/// Isolation is realised by the two WebKit primitives the JS policy maps onto:
/// an `isolated` process gets a fresh `WKProcessPool` (a distinct content
/// process); an `isolated` storage gets a non-persistent `WKWebsiteDataStore`
/// (its own cookies/localStorage/IndexedDB, nothing shared). `shared` reuses a
/// plugin-owned pool/the default store — never an implicit default, which is why
/// `createSurface` rejects when the policy fields are absent.
@objc(ElizaSurfaceManagerPlugin)
public class ElizaSurfaceManagerPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ElizaSurfaceManagerPlugin"
    public let jsName = "ElizaSurfaceManager"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "createSurface", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setBounds", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "navigate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "foregroundSurface", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "backgroundSurface", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "destroySurface", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "foregroundHost", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getSurfaceState", returnType: CAPPluginReturnPromise),
    ]

    private struct Surface {
        let webView: WKWebView
        let process: String
        let storage: String
    }

    private var surfaces: [String: Surface] = [:]
    // One plugin-owned pool for every `shared`-process surface — deliberate, so a
    // shared surface still never lands in the host's implicit default pool.
    private let sharedProcessPool = WKProcessPool()

    @objc func createSurface(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("createSurface requires an id")
            return
        }
        // Explicit-policy invariant: both axes must be stated. No default.
        guard let process = call.getString("process"),
              process == "isolated" || process == "shared" else {
            call.reject("createSurface requires an explicit process policy (isolated|shared)")
            return
        }
        guard let storage = call.getString("storage"),
              storage == "isolated" || storage == "shared" else {
            call.reject("createSurface requires an explicit storage policy (isolated|shared)")
            return
        }
        let urlString = call.getString("url")

        DispatchQueue.main.async {
            guard let hostView = self.bridge?.viewController?.view else {
                call.reject("no host view controller to attach the surface to")
                return
            }
            if self.surfaces[id] != nil {
                call.resolve()
                return
            }

            let config = WKWebViewConfiguration()
            // Fresh pool ⇒ distinct content process; shared ⇒ the plugin pool.
            config.processPool = process == "isolated" ? WKProcessPool() : self.sharedProcessPool
            // Non-persistent store ⇒ private, per-surface cookies/localStorage.
            config.websiteDataStore = storage == "isolated"
                ? WKWebsiteDataStore.nonPersistent()
                : WKWebsiteDataStore.default()

            let webView = WKWebView(frame: hostView.bounds, configuration: config)
            webView.isHidden = true
            hostView.addSubview(webView)

            if let urlString = urlString, let url = URL(string: urlString) {
                webView.load(URLRequest(url: url))
            }
            self.surfaces[id] = Surface(webView: webView, process: process, storage: storage)
            call.resolve()
        }
    }

    @objc func setBounds(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("setBounds requires an id")
            return
        }
        let x = call.getDouble("x") ?? 0
        let y = call.getDouble("y") ?? 0
        let width = call.getDouble("width") ?? 0
        let height = call.getDouble("height") ?? 0
        DispatchQueue.main.async {
            guard let surface = self.surfaces[id] else {
                call.reject("no surface \(id)")
                return
            }
            // CSS px map 1:1 to UIKit points, so no density conversion is needed.
            surface.webView.frame = CGRect(x: x, y: y, width: width, height: height)
            call.resolve()
        }
    }

    @objc func navigate(_ call: CAPPluginCall) {
        guard let id = call.getString("id"), let urlString = call.getString("url"),
              let url = URL(string: urlString) else {
            call.reject("navigate requires an id and a valid url")
            return
        }
        DispatchQueue.main.async {
            guard let surface = self.surfaces[id] else {
                call.reject("no surface \(id)")
                return
            }
            surface.webView.load(URLRequest(url: url))
            call.resolve()
        }
    }

    @objc func foregroundSurface(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("foregroundSurface requires an id")
            return
        }
        DispatchQueue.main.async {
            guard let surface = self.surfaces[id] else {
                call.reject("no surface \(id)")
                return
            }
            surface.webView.superview?.bringSubviewToFront(surface.webView)
            surface.webView.isHidden = false
            call.resolve()
        }
    }

    @objc func backgroundSurface(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("backgroundSurface requires an id")
            return
        }
        DispatchQueue.main.async {
            guard let surface = self.surfaces[id] else {
                call.reject("no surface \(id)")
                return
            }
            surface.webView.isHidden = true
            call.resolve()
        }
    }

    @objc func destroySurface(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("destroySurface requires an id")
            return
        }
        DispatchQueue.main.async {
            if let surface = self.surfaces.removeValue(forKey: id) {
                surface.webView.stopLoading()
                surface.webView.removeFromSuperview()
            }
            call.resolve()
        }
    }

    @objc func foregroundHost(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            for surface in self.surfaces.values {
                surface.webView.isHidden = true
            }
            call.resolve()
        }
    }

    @objc func getSurfaceState(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("getSurfaceState requires an id")
            return
        }
        DispatchQueue.main.async {
            guard let surface = self.surfaces[id] else {
                call.resolve([
                    "exists": false,
                    "foregrounded": false,
                    "currentUrl": NSNull(),
                    "process": NSNull(),
                    "storage": NSNull(),
                ])
                return
            }
            call.resolve([
                "exists": true,
                "foregrounded": !surface.webView.isHidden,
                "currentUrl": surface.webView.url?.absoluteString ?? NSNull(),
                "process": surface.process,
                "storage": surface.storage,
            ])
        }
    }
}
